### ISSUE TITLE
jiffy:encode leaks memory when encoding maps

### ISSUE THREAD

**Issue by scohen** [davisp/jiffy#112]
When encoding maps, jiffy leaks memory, specifically, memory in the port driver area. This memory is not reclaimed by the VM.

The following JSON file will exercise the problem. It's extremely sensitive, removing a single key will make it not happen any more.

```json
[
    {   
                "abcdef": true, 
                "abc_defghijkl_mnop_qrstuvwxyz": true, 
                "abcde_fg_hijklmnop": true, 
                "abcde_fg_hijklmn_op_qrstu": true, 
                "email_repins": true, 
                "email_shares": true, 
                "email_suggestions": true, 
                "email_updates": true, 
                "first_name": "abcd", 
                "gender": "male", 
                "hidden_for": 0,  
                "hidden_from_public": false, 
                "hide_from_news": false, 
                "id": 95983773219368241, 
                "image_url": "abcdefgh_1334246760", 
                "is_hidden_due_to_ato": false, 
                "is_write_banned": false, 
                "last_login": "2015-10-07T13:50:47.879", 
                "last_name": "abcdef", 
                "last_password_request": "2013-01-27T14:11:58.821", 
                "last_pin_like_time": "2015-09-27T03:07:05.263", 
                "locale": "en-US", 
                "needs_captcha": false, 
                "needs_reputation_nag": false, 
                "orientation": true, 
                "personalize_from_offsite_browsing": true, 
                "publish_facebook": false, 
                "publish_timeline": false, 
                "publish_twitter": false, 
                "redacted": false, 
                "session_valid_after": 1359295919, 
                "username": "abcdefg", 
                "voluntarily_deactivated": false
    }   
]
```

I've been testing with the following code (sorry about the elixir). The only dependency is the recon library. 
I'm using Poison to demonstrate that it's not decoding that causes the problem. The memory leak also happens if you do `:jiffy.encode(json, [:use_maps])`, which is elixir for `jiffy:encode(json, [use_maps])`.

```elixir
  def decode_and_encode(json_string) do
     json
     |> Poison.decode!
     |> :jiffy.encode
  end

  def dump_memory do 
    :recon_alloc.set_unit :megabyte
    :recon_alloc.memory :allocated_types
  end

  def test_memory_leak(num_iterations) do
    giant_file = File.read!("/path/to/memory_leak.json")
    (1..num_iterations) |> Enum.each fn _ ->
      decode_and_encode(giant_file)
    end
  end
```

In the elixir shell, running 
```elixir
test_memory_leak(100_000)
```

will cause  driver_alloc to steadily increase. The following output happens after 100,000 iterations (units are in megabytes). 
```elixir
[binary_alloc: 1.7827606201171875, driver_alloc: 21.532760620117188,
 eheap_alloc: 20.880416870117188, ets_alloc: 6.0327606201171875,
 fix_alloc: 3.0327606201171875, ll_alloc: 33.251441955566406,
 sl_alloc: 0.2827606201171875, std_alloc: 1.2827606201171875,
 temp_alloc: 1.1260299682617188]
```

Over time, this has caused our servers to run out of memory, we do a lot of JSON encoding and decoding. 

In all other aspects, Jiffy is awesome. Thanks!

**Comment by jparise** [comment]
It looks like the map iterator in `enc_map_to_ejson()` is never destroyed. Calls to [enif_map_iterator_create](http://www.erlang.org/doc/man/erl_nif.html#enif_map_iterator_create) must be paired with a call to [enif_map_iterator_destroy](http://www.erlang.org/doc/man/erl_nif.html#enif_map_iterator_destroy).

**Comment by scohen** [comment]
I can confirm that @jparise 's PR fixes the issue.

**Comment by scohen** [comment]
@davisp, any love? This is a fairly serious issue for anyone that does heavy encoding/decoding.

**Comment by davisp** [comment]
Whoops, totally dropped the ball on this one over the holiday weekend. I merged #113 and its in the 0.14.5 tag.

**Comment by scohen** [comment]
No worries; I figured as much. 
Thanks so much, we're very heavy users of Jiffy and it works very well ;)
